File: src/gui/Hyram.Units/Units.cs
/*
Copyright 2015-2022 National Technology & Engineering Solutions of Sandia, LLC (NTESS).
Under the terms of Contract DE-NA0003525 with NTESS, the U.S.Government retains certain
rights in this software.

You should have received a copy of the GNU General Public License along with
HyRAM+. If not, see https://www.gnu.org/licenses/.
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Serialization;

namespace SandiaNationalLaboratories.Hyram
{
    public enum UnitlessUnit { Unitless }

    public enum SpeedUnit { MetersPerSecond }

    public enum VolumetricFlowUnit { CubicMetersPerSecond }

    public enum DistanceUnit { Meter, Centimeter, Millimeter, Inch, Foot, Yard, Mile, Au }

    public enum TempUnit { Celsius, Fahrenheit, Kelvin }

    public enum PressureUnit { MPa, kPa, Pa, Psi, Atm, Bar, JoulePerCubicMeter }

    public enum AreaUnit { SqMeters, SqCm, SqMm, SqInch, SqFoot, SqYard }

    public enum DensityUnit { KilogramPerCubicMeter, GramPerCubicMeter, GramPerCubicCentimeter, MilligramPerLiter, OuncePerCubicFoot, OuncePerGallonUK, OuncePerGallonUS }

    public enum EnergyUnit { Joule, Kwh, Botu }

    public enum AngleUnit { Radians, Degrees }

    public enum MassUnit { Gram, Milligram, Centigram, Decigram, Dekagram, Hectogram, Kilogram, Megagram, Pound }
    
    public enum VolumeUnit
    {
        Liter,
        CubicCentimeter, CubicDecimeter, CubicDekameter, CubicFoot, CubicInch, CubicKilometer,
        CubicMeter, CubicMile, CubicMicrometer, CubicMillimeter, CubicYard, Deciliter,
        Dekaliter, Kiloliter, Megaliter, Microliter, Milliliter
    }

    public enum SpecificEnergyUnit { JoulePerKilogram, JoulePerGram, KiloJoulePerKilogram }

    // Manpower Time Unit isn't a general conversion enumeration.
    public enum ManpowerTimeUnit { Hour }

    public enum JulianTimeConversionUnit { Year, Day } // Year is 365.25

    public enum TimeUnit { Hour, Minute, Second, Millisecond }

    public enum FreqUnit { PerYear }

    public enum MassFlowUnit { KgPerSecond }


    public static class UnitParser
    {
        private static readonly Dictionary<string, SpecificEnergyUnit> MSpecificEnergyUnits = CreateUnitDictionary<SpecificEnergyUnit>();
        private static readonly Dictionary<string, DistanceUnit> MDistanceUnits = CreateUnitDictionary<DistanceUnit>();
        private static readonly Dictionary<string, MassUnit> MMassUnits = CreateUnitDictionary<MassUnit>();
        private static readonly Dictionary<string, MassFlowUnit> MMassFlowUnits = CreateUnitDictionary<MassFlowUnit>();
        private static readonly Dictionary<string, VolumeUnit> MVolumeUnits = CreateUnitDictionary<VolumeUnit>();
        private static readonly Dictionary<string, DensityUnit> MDensityUnits = CreateUnitDictionary<DensityUnit>();

        public static PressureUnit ParsePressureUnit(string unitName)
        {
            PressureUnit result;

            var unitNameUpper = unitName.Trim().ToUpper();
            switch (unitNameUpper)
            {
                case "MPA":
                    result = PressureUnit.MPa;
                    break;
                case "KPA":
                    result = PressureUnit.kPa;
                    break;
                case "PA":
                    result = PressureUnit.Pa;
                    break;
                case "PSIG":
                case "PSI":
                    result = PressureUnit.Psi;
                    break;
                case "ATM":
                    result = PressureUnit.Atm;
                    break;
                case "BAR":
                    result = PressureUnit.Bar;
                    break;
                case "JCM":
                    result = PressureUnit.JoulePerCubicMeter;
                    break;
                default:
                    throw new ArgumentException("Unit of " + unitName + " is not a recognized Pressure unit.");
            }

            return result;
        }

        public static TempUnit ParseTempUnit(string unitName)
        {
            TempUnit result;
            var unitNameUpper = unitName.Trim().ToUpper();
            switch (unitNameUpper)
            {
                case "CELSIUS":
                    result = TempUnit.Celsius;
                    break;
                case "FAHRENHEIT":
                    result = TempUnit.Fahrenheit;
                    break;
                case "KELVIN":
                    result = TempUnit.Kelvin;
                    break;
                default:
                    throw new ArgumentException("Unit of " + unitName + " is not a recognized Temperature unit.");
            }

            return result;
        }

        public static DensityUnit ParseDensityUnit(string unitName)
        {
            if (MDensityUnits.ContainsKey(unitName))
            {
                return MDensityUnits[unitName];
            }
            throw new ArgumentException("Unit of " + unitName + " is not a recognized density unit.");
        }

        public static VolumeUnit ParseVolumeUnit(string unitName)
        {
            if (MVolumeUnits.ContainsKey(unitName))
            {
                return MVolumeUnits[unitName];
            }
            throw new ArgumentException("Unit of " + unitName + " is not a recognized volume unit.");
        }


        public static MassUnit ParseMassUnit(string unitName)
        {
            if (MMassUnits.ContainsKey(unitName))
            {
                return MMassUnits[unitName];
            }
            throw new ArgumentException("Unit of " + unitName + " is not a recognized mass unit.");
        }

        public static MassFlowUnit ParseMassFlowUnit(string unitName)
        {
            if (MMassFlowUnits.ContainsKey(unitName))
            {
                return MMassFlowUnits[unitName];
            }
            throw new ArgumentException("Unit of " + unitName + " is not a recognized mass unit.");
        }


        public static DistanceUnit ParseDistanceUnit(string unitName)
        {
            if (MDistanceUnits.ContainsKey(unitName))
            {
                return MDistanceUnits[unitName];
            }
            throw new ArgumentException("Unit of " + unitName + " is not a recognized distance unit.");
        }

        public static SpecificEnergyUnit ParseSpecificEnergyUnit(string unitName)
        {
            if (MSpecificEnergyUnits.ContainsKey(unitName))
            {
                return MSpecificEnergyUnits[unitName];
            }
            throw new ArgumentException("Unit of " + unitName + " is not a recognized specific energy unit.");
        }



        private static IEnumerable<T> GetEnumPossibleValues<T>()
        {
            return Enum.GetValues(typeof(T)).Cast<T>();
        }

        private static Dictionary<string, SpecificEnergyUnit> CreateSpecificEnergyDictionary()
        {
            var result = new Dictionary<string, SpecificEnergyUnit>();
            var possibleValues =
                (SpecificEnergyUnit[]) GetEnumPossibleValues<SpecificEnergyUnit>();
            foreach (var thisValue in possibleValues) result.Add(thisValue.ToString(), thisValue);

            return result;
        }

        private static Dictionary<string, DistanceUnit> CreateDistanceUnitsDictionary()
        {
            var result = new Dictionary<string, DistanceUnit>();
            var possibleValues =
                (DistanceUnit[]) GetEnumPossibleValues<DistanceUnit>();
            foreach (var thisValue in possibleValues) result.Add(thisValue.ToString(), thisValue);

            return result;
        }

        private static Dictionary<string, MassUnit> CreateMassUnitsParsingDictionary()
        {
            var result = new Dictionary<string, MassUnit>();
            var possibleValues = (MassUnit[]) GetEnumPossibleValues<MassUnit>();
            foreach (var thisValue in possibleValues) result.Add(thisValue.ToString(), thisValue);

            return result;
        }

        private static Dictionary<string, VolumeUnit> CreateVolumeUnitsParsingDictionary()
        {
            var result = new Dictionary<string, VolumeUnit>();
            var possibleValues = (VolumeUnit[]) GetEnumPossibleValues<VolumeUnit>();
            foreach (var thisValue in possibleValues) result.Add(thisValue.ToString(), thisValue);

            return result;
        }

        private static Dictionary<string, DensityUnit> CreateDensityUnitsParsingDictionary()
        {
            var result = new Dictionary<string, DensityUnit>();
            var possibleValues = (DensityUnit[]) GetEnumPossibleValues<DensityUnit>();
            foreach (var thisValue in possibleValues) result.Add(thisValue.ToString(), thisValue);
            return result;
        }

        private static Dictionary<string, TUnit> CreateUnitDictionary<TUnit>()
        {
            var result = new Dictionary<string, TUnit>();
            var possibleValues = (TUnit[]) GetEnumPossibleValues<TUnit>();
            foreach (var thisValue in possibleValues) result.Add(thisValue.ToString(), thisValue);
            return result;
        }

    }

}

File: gui/src/hyramgui/models/units.py
"""
Copyright 2015-2025 National Technology & Engineering Solutions of Sandia, LLC (NTESS).
Under the terms of Contract DE-NA0003525 with NTESS, the U.S. Government retains certain rights in this software.

You should have received a copy of the GNU General Public License along with HyRAM+.
If not, see https://www.gnu.org/licenses/.

"""
from hyramgui.hygu.utils.units_of_measurement import UnitType
import scipy.constants as scp


class Pressure(UnitType):
    """Pressure unit options. """
    label = 'pres'
    unit_data = {'pa': 1, 'kpa': scp.kilo, 'mpa': scp.mega, 'psi': scp.psi, 'atm': scp.atm, 'bar': scp.bar}
    display_units = ['Pa', 'kPa', 'MPa', 'psi', 'atm', 'bar']
    pa = 'pa'
    kpa = 'kpa'
    mpa = 'mpa'
    psi = 'psi'
    atm = 'atm'
    bar = 'bar'
    std_unit = 'pa'


class Time(UnitType):
    label = 'time'
    unit_data = {'hr': scp.hour, 'min': scp.minute, 'sec': 1, 'ms': scp.milli}
    display_units = ['hr', 'min', 's', 'ms']
    hr = 'hr'
    min = 'min'
    sec = 'sec'
    ms = 'ms'
    std_unit = 'sec'


class Area(UnitType):
    label = 'area'
    unit_data = {'m2': 1, 'cm2': scp.centi**2, 'mm2': scp.milli**2, 'in2': scp.inch**2, 'ft2': scp.foot**2, 'yd2': scp.yard**2}
    display_units = ['m<sup>2</sup>', 'cm<sup>2</sup>', 'mm<sup>2</sup>', 'in<sup>2</sup>', 'ft<sup>2</sup>', 'yd<sup>2</sup>']
    m2 = 'm2'
    cm2 = 'cm2'
    mm2 = 'mm2'
    in2 = 'in2'
    ft2 = 'ft2'
    yd2 = 'yd2'
    std_unit = 'm2'


class Volume(UnitType):
    label = 'volume'
    unit_data = {'l': scp.liter, 'm3': 1, 'mm3': scp.milli**3, 'cm3': scp.milli**2, 'ml': scp.milli**2,
                 'in3': scp.inch**3, 'ft3': scp.foot**3}
    display_units = ['L', 'm<sup>3</sup>', 'mm<sup>3</sup>', 'cm<sup>3</sup>', 'mL', 'in<sup>3</sup>', 'ft<sup>3</sup>']
    l = 'l'
    m3 = 'm3'
    mm3 = 'mm3'
    cm3 = 'cm3'
    ml = 'ml'
    in3 = 'in3'
    ft3 = 'ft3'
    std_unit = 'm3'


class Density(UnitType):
    label = 'density'
    unit_data = {'kgpm3': 1, 'gpm3': 1/scp.kilo, 'gpcm3': 1/scp.centi**3/scp.kilo, 'kgpl': 1/scp.liter, 'gpl': 1/scp.liter/scp.kilo, 'lbpft3': scp.pound/scp.foot**3}
    display_units = ['kg/m<sup>3</sup>', 'g/m<sup>3</sup>', 'g/cm<sup>3</sup>', 'kg/L', 'g/l', 'lb/ft<sup>3</sup>']
    std_unit = 'kgpm3'
    kgpm3 = 'kgpm3'
    gpm3 = 'gpm3'
    gpcm3 = 'gpcm3'
    kgpl = 'kgpl'
    gpl = 'gpl'
    lbpft3 = 'lbpft3'


class Mass(UnitType):
    label = 'mass'
    unit_data = {'mg': scp.milli/scp.kilo, 'g': 1/scp.kilo, 'kg': 1, 'lb': scp.pound}
    display_units = ['mg', 'g', 'kg', 'lb']
    std_unit = 'kg'
    mg = 'mg'
    g = 'g'
    kg = 'kg'
    lb = 'lb'
